enhancements/kube-apiserver/stability-point-to-point-network-check.md

```diff
@@ -0,0 +1,447 @@
+* Limiting scope to DNS lookup and TCP connect to target endpoint. REST client can fail for a variety of other reasons.
+* Providing insights into target endpoints that are not accessed a REST client.
+* Detecting and reporting issues independently of monitoring, which might itself not be available.
+* Providing insights during a rollout.
+* Providing status that can be consumed by an operator.
+
+### Goals
+
+1. be independent of SDN, cloud, or cluster ingress
+2. Identify connection failure between each kas and each oas endpoint
+3. Identify connection failure between each kas and each etcd
+4. Identify connection failure between each kas and oas service IP
+5. Identify connection failure between each oas and each kas host IP
+6. Identify connection failure between each oas and each kas endpoint IP
+7. Identify connection failure between each oas and kas service IP
+8. Identify connection failure between each oas and each etcd
+
+### Non-Goals
+
+## Proposal
+
+### 1. One instance of a new `PodNetworkConnectivityCheck` CR per point to point connection.
+
+A point to point connection is between a source pod and a target endpoint.
+
+Component operators will create a `PodNetworkConnectivityCheck` resource per point-to-point connection. If an operator 
+is not managing the source pod directly, but via a workload controller (such as via a deployment or daemonset), the 
+operator will observe the workload resources to determine the list of pods and create a `PodNetworkConnectivityCheck` 
+resource for each target endpoint per pod. Operators that directly manage a source pod (such as operators that manage 
+static-pods) will create `PodNetworkConnectivityCheck` resources for each target endpoint when creating the source pod.
+
+### 2. Detailed status, probably some kind of latency information.
+
+The `PodNetworkConnectivityCheck` will maintain a log of the actions performed during a check, recording if the actions
+were successful, a reason for success or failure, latency of performing the action and some human readable message.
+
+**LogEntry:**
+```go
+// LogEntry records events
+type LogEntry struct {
+	// Start time of check action.
+	Start metav1.Time `json:"time,omitempty"`
+	// Status indicates if the action performed by the check was successful or not.
+	Success bool `json:"status"`
+	// Reason for status in a machine readable format.
+	Reason LogEntryReason `json:"reason"`
+	// Message explaining status in a human readable format.
+	Message string `json:"message"`
+	// Latency records how long the action mentioned in the entry took.
+	Latency time.Duration `json:"latency"`
+}
+``` 
+
+The start and end time of detected outages will be logged separately from the individual check actions.
+
+### 3. It should be possible to clean up garbage after a couple hours of inactivity.
+
+A controller to prune `PodNetworkConnectivityCheck` resources, will be provided for inclusion into source pod operators.
```

Comment:
a pruner configured via command line args seems to be much simpler and easier to maintain.